src/main/scala/calc/Compiler.scala

```diff
@@ -74,8 +98,60 @@ object Compiler {
       case Literal(value) =>
         irt.DoubleLiteral(value)
 
-      case _ =>
-        throw new Exception(
+      case BinaryOp(op, lhs, rhs) => {
+        val jsop = op match {
+          case "+" => irt.BinaryOp.Double_+
+          case "-" => irt.BinaryOp.Double_-
+          case "*" => irt.BinaryOp.Double_*
+          case "/" => irt.BinaryOp.Double_/
+        }
+        irt.BinaryOp(jsop, compileExpr(lhs)(env, typeEnv), compileExpr(rhs)(env, typeEnv))
+      }
+
+      case Let(ident, value, body) => {
+        val jsV = compileExpr(value)(env, typeEnv)
+        val jsIdent = irt.Ident(ident.name)
+        val jsId = irt.VarDef(jsIdent, jsV.tpe, false, jsV)
+        irt.Block(List(jsId, compileExpr(body)(env+(ident.name -> jsV), typeEnv+(ident.name -> jsV.tpe))))
+      }
+
+      case Ident(n) => {
+        try {
+          typeEnv(n) match {
+            case irtpe.DoubleType => irt.Unbox(irt.VarRef(irt.Ident(n))(typeEnv(n)),'D')
+            case _ => irt.VarRef(irt.Ident(n))(typeEnv(n))
+          }
+        }catch{
+          case e:Exception => irt.VarRef(irt.Ident(n)) (irtpe.DoubleType)
+        }
+      }
+
+      case If(cond, thenp, elsep) => {
+        val jsThenp = compileExpr(thenp)(env, typeEnv)
+        val jsElsep = compileExpr(elsep)(env, typeEnv)
+        val jsCond = compileExpr(cond)(env, typeEnv)
+        val finalCond = irt.BinaryOp(irt.BinaryOp.!==, irt.Unbox(jsCond,'D'), irt.DoubleLiteral(0.0))
+        if (jsElsep.tpe != jsElsep.tpe){
+          throw new Exception(s"The then and also phase in the conditional statement must bse same type")
+        }
+        irt.If(finalCond, jsThenp, jsElsep)(jsThenp.tpe)
+      }
+
+      case Closure(idList, body) => {
+        val captureList = findCapture(tree).toList
+        val captureParam = captureList.map(s => irt.ParamDef(irt.Ident(s), typeEnv(s), false, false))
+        val params = idList.map(s => irt.ParamDef(irt.Ident(s.name), irtpe.AnyType, false, false))
+        val newTypeEnv = idList.foldLeft(typeEnv) ((env:Map[String,irtpe.Type], s:Ident) => env+(s.name -> irtpe.DoubleType))
```

Comment:
I think you're mistake is here. You create params with `s.name` and type `any`, but you add in your `envType` that `s.name` has type `double`.

Instead, you should declare params with other (generated) names, of type `any`, and add `VarDef`s at the beginning of the method to unbox the params in local variable of name `s.name` and type `double`. Then it's ok to use `s.name` of type `double` in the environment.